### PR TITLE
compact rules which has the same ResourceName

### PR DESCRIPTION
**Pull request by xilabao**
**What this PR does / why we need it**:
make compact rule more elegant.
Before
```
# ./cluster/kubectl.sh describe clusterrole system:kube-scheduler
Name:		system:kube-scheduler
Labels:		kubernetes.io/bootstrapping=rbac-defaults
Annotations:	rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources			Non-Resource URLs	Resource Names		Verbs
  ---------			-----------------	--------------		-----
  bindings			[]			[]			[create]
  endpoints			[]			[]			[create]
  endpoints			[]			[kube-scheduler]	[delete]
  endpoints			[]			[kube-scheduler]	[get]
  endpoints			[]			[kube-scheduler]	[patch]
  endpoints			[]			[kube-scheduler]	[update]
  events			[]			[]			[create patch update]
  nodes				[]			[]			[get list watch]
  persistentvolumeclaims	[]			[]			[get list watch]
  persistentvolumes		[]			[]			[get list watch]
  pods				[]			[]			[get list watch]
  pods/binding			[]			[]			[create]
  pods/status			[]			[]			[update]
  replicasets.extensions	[]			[]			[get list watch]
  replicationcontrollers	[]			[]			[get list watch]
  services			[]			[]			[get list watch]
  statefulsets.apps		[]			[]			[get list watch]
```
After
```
# ./cluster/kubectl.sh describe clusterrole system:kube-scheduler
Name:		system:kube-scheduler
Labels:		kubernetes.io/bootstrapping=rbac-defaults
Annotations:	rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources			Non-Resource URLs	Resource Names		Verbs
  ---------			-----------------	--------------		-----
  bindings			[]			[]			[create]
  endpoints			[]			[]			[create]
  endpoints			[]			[kube-scheduler]	[delete get patch update]
  events			[]			[]			[create patch update]
  nodes				[]			[]			[get list watch]
  persistentvolumeclaims	[]			[]			[get list watch]
  persistentvolumes		[]			[]			[get list watch]
  pods				[]			[]			[get list watch]
  pods/binding			[]			[]			[create]
  pods/status			[]			[]			[update]
  replicasets.extensions	[]			[]			[get list watch]
  replicationcontrollers	[]			[]			[get list watch]
  services			[]			[]			[get list watch]
  statefulsets.apps		[]			[]			[get list watch]
```
**Which issue this PR fixes**:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by xilabao** [comment]
@k8s-bot tell me a joke

**Comment by liggitt** [comment]
@k8s-bot ok to test

**Comment by liggitt** [comment]
@kubernetes/sig-auth-pr-reviews

**Comment by xilabao** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by xilabao** [comment]
@liggitt ptal, thanks

File: pkg/registry/rbac/validation/policy_compact.go [hunk 1]
```diff
@@ -66,24 +71,30 @@ func CompactRules(rules []rbac.PolicyRule) ([]rbac.PolicyRule, error) {
...
-// isSimpleResourceRule returns true if the given rule contains verbs, a single resource, a single API group, and no other values
-func isSimpleResourceRule(rule *rbac.PolicyRule) (schema.GroupResource, bool) {
-	resource := schema.GroupResource{}
+// isSimpleResourceRule returns true if the given rule contains verbs, a single resource, a single API group, at most one Resource Name, and no other values
+func isSimpleResourceRule(rule *rbac.PolicyRule) (simpleResource, bool) {
+	resource := simpleResource{}
 
 	// If we have "complex" rule attributes, return early without allocations or expensive comparisons
-	if len(rule.ResourceNames) > 0 || len(rule.NonResourceURLs) > 0 {
+	if len(rule.ResourceNames) > 1 || len(rule.NonResourceURLs) > 0 {
```
Comment (liggitt) [inline pkg/registry/rbac/validation/policy_compact.go]: this equates `resourceNames:[]` and `resourceNames:[""]`, which is not accurate

Comment (xilabao) [inline pkg/registry/rbac/validation/policy_compact.go]: fixed. PTAL @liggitt

**Review by liggitt — COMMENTED** [review]
/lgtm

**Comment by xilabao** [comment]
@liggitt "approve" seems not success, could you approve once again? sorry for bothering you.

**Pull request merged**